File: kor-travel-geo-ui/lib/file-inventory.ts
import type { components } from "@/types/api.gen";

/** T-283 파일 관리 — 백엔드 통합 파일 인벤토리 계약 타입 (생성 타입 alias). */
export type FileInventoryItem = components["schemas"]["FileInventoryItem"];
export type FileInventoryPage = components["schemas"]["FileInventoryPage"];
export type FileInventorySummary = components["schemas"]["FileInventorySummary"];
export type FileInventorySourceDetail = components["schemas"]["FileInventorySourceDetail"];
export type FileInventoryUsage = components["schemas"]["FileInventoryUsage"];
export type FileInventoryFileInfo = components["schemas"]["FileInventoryFileInfo"];
export type FileInventorySessionInfo = components["schemas"]["FileInventorySessionInfo"];
export type FileInventoryIssue = components["schemas"]["FileInventoryIssue"];

export type FileInventoryKind = FileInventoryItem["file_kind"];

export const fileInventoryPaths = {
  list: (params: {
    kind?: string;
    category?: string;
    lifecycle?: string;
    temporaryOnly?: boolean;
    limit?: number;
  }) => {
    const query = new URLSearchParams();
    if (params.kind && params.kind !== "all") query.set("kind", params.kind);
    if (params.category) query.set("category", params.category);
    if (params.lifecycle) query.set("lifecycle", params.lifecycle);
    if (params.temporaryOnly) query.set("temporary_only", "true");
    if (params.limit) query.set("limit", String(params.limit));
    const text = query.toString();
    return `/admin/storage/files${text ? `?${text}` : ""}`;
  },
  sourceGroupDetail: (groupId: string) =>
    `/admin/storage/files/source-groups/${encodeURIComponent(groupId)}`
};

export const fileKindLabels: Record<string, string> = {
  source_group: "원천 파일",
  artifact: "백업/산출물",
  orphan_object: "저장소 객체"
};

export interface LifecycleMeta {
  label: string;
  tone: "ok" | "warn" | "error" | "info" | "neutral" | "brand";
  /** HelpTip에 쓰는 한 줄 설명. */
  hint: string;
}

/** lifecycle 버킷의 한국어 라벨·배지 톤·설명 (백엔드 core.file_inventory와 동기). */
const lifecycleMeta: Record<string, LifecycleMeta> = {
  serving: { label: "서빙 사용 중", tone: "ok", hint: "활성 매칭 세트에 포함되어 운영 DB 구성에 사용 중입니다." },
  staging: { label: "구성 대기", tone: "info", hint: "매칭 세트에 포함되어 있지만 아직 활성 세트는 아닙니다." },
  idle: { label: "미사용", tone: "neutral", hint: "등록은 됐지만 어떤 매칭 세트도 참조하지 않습니다." },
  in_progress: { label: "진행 중", tone: "warn", hint: "업로드 또는 검증이 진행 중인 임시 상태입니다." },
  unregistered: { label: "미등록", tone: "warn", hint: "저장은 됐지만 등록 기한이 지나 임시 상태로 남아 있습니다." },
  quarantined: { label: "격리", tone: "error", hint: "정합성 문제로 격리된 상태입니다." },
  missing: { label: "유실", tone: "error", hint: "DB 기록은 있으나 저장소에서 객체를 찾지 못했습니다." },
  soft_deleted: { label: "삭제 예정", tone: "neutral", hint: "soft-delete 상태 — 복원할 수 있습니다." },
  hard_deleted: { label: "영구 삭제됨", tone: "neutral", hint: "저장소에서 영구 삭제됐습니다." },
  delete_failed: { label: "삭제 실패", tone: "error", hint: "삭제 시도가 실패해 정리가 필요합니다." },
  available: { label: "보관 중", tone: "ok", hint: "사용 가능한 상태로 보관 중입니다." },
  creating: { label: "생성 중", tone: "warn", hint: "생성 작업이 진행 중입니다." },
  failed: { label: "실패", tone: "error", hint: "생성 작업이 실패했습니다 — 정리 대상입니다." },
  expired: { label: "만료", tone: "neutral", hint: "보존 기간이 지나 만료됐습니다." },
  deleted: { label: "삭제됨", tone: "neutral", hint: "삭제 처리된 기록입니다." },
  orphan: { label: "미등록 객체", tone: "warn", hint: "저장소에만 존재하고 DB에 등록되지 않은 객체입니다 (RustFS 정합성 탭에서 정리)." }
};

export function lifecycleOf(value: string): LifecycleMeta {
  return lifecycleMeta[value] ?? { label: value, tone: "neutral", hint: value };
}

File: src/kortravelgeo/core/file_inventory.py
"""File-inventory lifecycle derivation (T-283 파일 관리).

Pure decision logic for the unified stored-file inventory (`/v1/admin/storage/files`):
given facts about a source file group (or a backup/ops artifact), derive a single
`lifecycle` bucket plus the `in_use` / `temporary` flags the admin UI filters on.
DB-free by design — the repository maps rows to these facts and the API composes
the results (mirrors ``core.source_janitor``).
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal

from kortravelgeo.dto.file_inventory import FileInventoryItem, FileInventorySummary
from kortravelgeo.dto.source import TERMINAL_UPLOAD_SESSION_STATES

#: Unified lifecycle buckets across source groups / artifacts / orphan objects.
FileLifecycle = Literal[
    # source groups
    "serving",  # 활성 매칭 세트에 포함 — 서빙 구성에 사용 중
    "staging",  # 매칭 세트에는 포함되나 활성 세트는 아님
    "idle",  # 등록·검증 완료지만 어떤 매칭 세트도 참조하지 않음
    "in_progress",  # 업로드/검증 진행 중
    "unregistered",  # 저장은 됐지만 등록 기한이 지난 임시 상태
    "quarantined",
    "missing",
    "soft_deleted",
    "hard_deleted",
    "delete_failed",
    # artifacts
    "available",
    "creating",
    "failed",
    "expired",
    "deleted",
    # reconcile orphan objects (RustFS에만 존재)
    "orphan",
]

_GROUP_TERMINAL_STATES: frozenset[str] = frozenset(
    {"quarantined", "missing", "soft_deleted", "hard_deleted", "delete_failed"}
)

#: lifecycle 값 중 "임시/정리 대상" 성격 — UI의 임시 필터 기준.
TEMPORARY_LIFECYCLES: frozenset[str] = frozenset(
    {"in_progress", "unregistered", "creating", "failed", "orphan", "delete_failed"}
)


@dataclass(frozen=True, slots=True)
class SourceGroupFacts:
    """Row-level facts needed to classify one ``ops.source_file_groups`` entry."""

    group_state: str
    session_state: str | None
    registered_at_present: bool
    active_match_set_id: str | None
    match_set_count: int
    candidate_match_set_count: int


@dataclass(frozen=True, slots=True)
class LifecycleVerdict:
    lifecycle: str
    in_use: bool
    temporary: bool


def derive_source_group_lifecycle(facts: SourceGroupFacts) -> LifecycleVerdict:
    """Classify a source file group into one lifecycle bucket.

    우선순위: 그룹 터미널 상태(격리/삭제 등) > 진행 중 세션 > 등록 만료 >
    매칭 세트 사용 여부. ``in_use``는 활성 매칭 세트 포함 여부만 본다(서빙 기준).
    """

    in_use = facts.active_match_set_id is not None

    if facts.group_state in _GROUP_TERMINAL_STATES:
        return LifecycleVerdict(
            facts.group_state,
            in_use=False,
            temporary=facts.group_state == "delete_failed",
        )

    session_active = (
        facts.session_state is not None
        and facts.session_state not in TERMINAL_UPLOAD_SESSION_STATES
    )
    if session_active or facts.group_state == "validating":
        return LifecycleVerdict("in_progress", in_use=in_use, temporary=True)

    if facts.session_state == "registration_expired" and not facts.registered_at_present:
        return LifecycleVerdict("unregistered", in_use=False, temporary=True)

    if in_use:
        return LifecycleVerdict("serving", in_use=True, temporary=False)
    if facts.candidate_match_set_count > 0:
        return LifecycleVerdict("staging", in_use=False, temporary=False)
    return LifecycleVerdict("idle", in_use=False, temporary=False)


def derive_artifact_lifecycle(state: str) -> LifecycleVerdict:
    """Classify an ``ops.artifacts`` row (backup 등 산출물 파일)."""

    if state == "creating":
        return LifecycleVerdict("creating", in_use=False, temporary=True)
    if state == "failed":
        return LifecycleVerdict("failed", in_use=False, temporary=True)
    if state in {"expired", "deleted"}:
        return LifecycleVerdict(state, in_use=False, temporary=False)
    return LifecycleVerdict("available", in_use=False, temporary=False)


def orphan_lifecycle() -> LifecycleVerdict:
    """RustFS에는 있으나 DB 레지스트리에 없는 객체 (reconcile open issue)."""

    return LifecycleVerdict("orphan", in_use=False, temporary=True)


def build_inventory_summary(items: Sequence[FileInventoryItem]) -> FileInventorySummary:
    """목록에서 UI 상단 요약(개수·용량·사용/임시/이슈 카운트)을 집계한다."""

    by_lifecycle: dict[str, int] = {}
    by_kind: dict[str, int] = {}
    total_bytes = 0
    in_use_count = 0
    temporary_count = 0
    open_issue_count = 0
    for item in items:
        by_lifecycle[item.lifecycle] = by_lifecycle.get(item.lifecycle, 0) + 1
        by_kind[item.file_kind] = by_kind.get(item.file_kind, 0) + 1
        total_bytes += item.size_bytes or 0
        if item.in_use:
            in_use_count += 1
        if item.temporary:
            temporary_count += 1
        open_issue_count += item.open_issue_count
    return FileInventorySummary(
        total_count=len(items),
        total_bytes=total_bytes,
        in_use_count=in_use_count,
        temporary_count=temporary_count,
        open_issue_count=open_issue_count,
        by_lifecycle=by_lifecycle,
        by_kind=by_kind,
    )
